cql3: detect and handle int overflow in aggregate functions

bhalevy opened the pull request:
Fix overflow handling in sum() and avg().

sum:
- aggregated into __int128
- detect overflow when computing result
  - and log a warning if found

avg:
- fix division function to divide the accumulator type _sum (__int128 for integers) by _count

- Add unit tests for both cases

Test:
- manual test against Cassandra 3.11.3 to make sure the results in the scylla unit test agree with it.
- unit(dev), cql_query_test(debug)

Fixes #5536

espindola commented on cql3/functions/functions.cc:
```diff
@@ -684,6 +684,20 @@ function_call::raw::to_string() const {
+void impl_sum_function_for<Type>::validate_sum() {
+    Type short_sum = static_cast<Type>(_sum);
+    if (static_cast<accumulator_type>(short_sum) != _sum) {
+        log.warn(format("SUM overflow: 0x{:08x}_{:08x}_{:08x}_{:08x} -> {}",
```
My main question is: Do we need this? Does cassandra produce the same warning?

If in CQL signed integer arithmetic is defined to be modular, why not just use std::make_unsigned to select the type and skip the warning?

bhalevy replied:
Regarding AVG - this agrees with what Cassandra produces (at least for the `int` type, not sure about the `bigint` type, but I'd consider it a bug if the internal accumulator overflows since the average of data of type T must always fit in T)

I can't say that SUM is "defined to be modular", but it behaves this way in Cassandra.
From the user's perspective I think it's bad to lose information about overflow.
Id we suppress any warning then I'd be happy if we add a way in CQL to declare a type type for the SUM that is different than the column type, e.g. `SELECT SUM(val) FROM cf AS bigint;`

bhalevy replied:
@avi / @slivne / @tzach / @nyh  what's you take on the above?

bhalevy replied:
Summoning also @GalitElad

avikivity replied:
Better to throw an exception (and fail the query). There's no point in returning a modulo arithmetic sum.

The user can widen the calculation by using `SELECT SUM(val AS bigint) FROM tab`.

What do other databases do? Perhaps it makes sense for SUM to auto-widen, at least up to bigint.

bhalevy replied:
I'm all for either auto-widen the result (easy), or throw an exception if it doesn't fit (not very hard either)

FWIW, Cassandra decided on silently truncating the result.
See https://issues.apache.org/jira/browse/CASSANDRA-9674

avikivity replied:
Auto-widen is not so easy, because it changes the prepared statement metadata. So we have to introduce a cluster feature, and hash the feature into prepared statement ID so we don't mix an old prepared statement with a new one.

bhalevy replied:
Though I'm not sure how would the clients handle possible overflow in the result.

I suggest that as a first step we throw an error since, as @avikivity mentioned, the user/app can promote the input to safely retrieve a larger result.

Then, as a follow-up, consider auto-promotion of the result type if feasible.

avikivity replied:
But let's collect information about what other databases do.

espindola commented on cql3/functions/functions.cc:
```diff
@@ -684,6 +684,20 @@ function_call::raw::to_string() const {
+namespace aggregate_fcts {
+
+template <typename Type>
+void impl_sum_function_for<Type>::validate_sum() {
```
We should move all of this code to the .cc file, but having a template declaration in a header and the definition in a .cc file is asking for trouble. Please keep this in the .hh for now.

I will send a RFC patch doing the move and rebase on top of this once it is in.

bhalevy replied:
makes sense

espindola commented on test/boost/cql_query_test.cc:
```diff
@@ -4313,3 +4313,55 @@ SEASTAR_TEST_CASE(test_rf_expand) {
+    });
+}
+
+SEASTAR_TEST_CASE(test_int_sum_overflow) {
```
This tests already pass is non debug builds, which suggests that std::make_unsigned is the way to go.

bhalevy replied:
I agree from a programming perspective, but not sure if it's the right thing to do from the usability perspective.

slivne commented:
I don;t think we can auto widen the result - since the user can be
expecting a specific type returned and as such may have setup his
code accordingly. So we should either throw or truncate.

On Wed, Jan 1, 2020 at 5:31 PM Avi Kivity <notifications@github.com> wrote:

> *@avikivity* commented on this pull request.
> ------------------------------
>
> In cql3/functions/functions.cc
> <https://github.com/scylladb/scylla/pull/5537#discussion_r362329285>:
>
> > @@ -684,6 +684,20 @@ function_call::raw::to_string() const {
>      return format("{}({})", _name, join(", ", _terms));
>  }
>
> +namespace aggregate_fcts {
> +
> +template <typename Type>
> +void impl_sum_function_for<Type>::validate_sum() {
> +    Type short_sum = static_cast<Type>(_sum);
> +    if (static_cast<accumulator_type>(short_sum) != _sum) {
> +        log.warn(format("SUM overflow: 0x{:08x}_{:08x}_{:08x}_{:08x} -> {}",
>
> Auto-widen is not so easy, because it changes the prepared statement
> metadata. So we have to introduce a cluster feature, and hash the feature
> into prepared statement ID so we don't mix an old prepared statement with a
> new one.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/5537?email_source=notifications&email_token=AA2OCCEGN32LD3QN7TBKZRTQ3SZNTA5CNFSM4KBWY6J2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQO335I#discussion_r362329285>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA2OCCD4GP4CDSYT7RXGQEDQ3SZNTANCNFSM4KBWY6JQ>
> .
>

bhalevy commented:
In v3:
- Throw an exception on sum() overflow
- Added more unit tests for sum overflow and overflow-prevention by casting of input to bigint.

avikivity commented on exceptions/exceptions.hh:
```diff
@@ -203,6 +204,12 @@ struct read_failure_exception : public request_failure_exception {
+class overflow_error_exception: public cassandra_exception {
+public:
+    overflow_error_exception(sstring msg) noexcept
+        : cassandra_exception(exception_code::UNAUTHORIZED, std::move(msg)) {}
```
UNAUTHORIZED?

bhalevy replied:
sorry, not sure how this came to be this way.

avikivity commented on exceptions/exceptions.hh:
```diff
@@ -65,6 +65,7 @@ enum class exception_code : int32_t {
     WRITE_FAILURE   = 0x1500,
     READ_TIMEOUT    = 0x1200,
     READ_FAILURE    = 0x1300,
+    OVERFLOW_ERROR  = 0x1600,
```
I see this in Cassandra source:

```
src/java/org/apache/cassandra/exceptions/ExceptionCode.java:    CDC_WRITE_FAILURE   (0x1600),
```

Let's allocate our errors in the range 0x1aff-0x1a01 (backwards) to minimize the chance of collision.

bhalevy replied:
sure

bhalevy replied:
@avi, why `0x1aff-1a01` rather than `0x1fff-0x1f01`?

avikivity replied:
0x1fff is too special, somebody else might want to use it. 0x1aff is boring and no one would ever pick it on purpose.

bhalevy commented:
In v4:
- exceptions: define additional cassandra CQL exceptions codes
- Fixed overflow_error_exception code
- Define new OVERFLOW_ERROR exception code as 0x1fff

espindola commented on exceptions/exceptions.hh:
```diff
@@ -65,6 +65,8 @@ enum class exception_code : int32_t {
     WRITE_FAILURE   = 0x1500,
     READ_TIMEOUT    = 0x1200,
     READ_FAILURE    = 0x1300,
+    FUNCTION_FAILURE= 0x1400,
```
Nit for a follow-up: please sort.

bhalevy replied:
6ef26a4a4a6de3b0b46bb0b08b083761f90653ee added `WRITE_FAILURE`  disordered so it be close to WRITE_TIMEOUT.
But given the additional values sorting all of them does make more sense.

espindola commented on exceptions/exceptions.hh:
```diff
@@ -68,6 +68,10 @@ enum class exception_code : int32_t {
 
+    // Scylla-specific codes
+    // Allocated backwards from 0x1aff-0x1a01 to minimize the chance of collision with Cassandra.
+    OVERFLOW_ERROR  = 0x1aff,
```
Do we have good enough relations with cassandra developers to propose reserving this code on their end? They don't have to implement the exception now, but if they ever decide to we would at least be compatible.

bhalevy replied:
I definitely don't have these relations.
We can try sending a patch, but I doubt they'll accept it with no implemented use of this value.

avikivity replied:
They won't, they made it clear in another occasion.

espindola commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -75,15 +76,66 @@ class count_rows_function final : public native_aggregate_function {
     }
 };
 
+// We need a wider accumulator for sum and average,
```
The example of 2147483647 + 1 -1 is interesting. With this implementation we promise that we support intermediate values overflowing, as long as the final value is in range.

If we don't want to support that we could use  __builtin_add_overflow. I am OK with both, just checking.

bhalevy replied:
We want to support intermediate overflow to minimize user disruption.

avikivity replied:
Certainly it's important for average. avg(tinyint) always fits in a tinyint, but the intermediate accumulator can easily overflow.

For sum() it's not important, but we have it for free so why not.

espindola commented on test/boost/cql_query_test.cc:
```diff
@@ -4408,3 +4408,28 @@ SEASTAR_TEST_CASE(test_int_sum_with_cast) {
     });
 }
+
+SEASTAR_TEST_CASE(test_bigint_avg) {
```
This test is already passing. Did you intent to use 9223372036854775807 as the constants?

bhalevy replied:
yes, probably copy/paste error. good catch.
I'll fix that and resend a new version (including sorting of the error codes).

bhalevy commented:
In v5:
- Added "exceptions: sort exception_code definitions"
- Fixed test_bigint_avg to induce overflow (pre-fix) using max bigint (9223372036854775807)
